eng/scripts/run_tests.ps1

```diff
@@ -0,0 +1,86 @@
+Param(
+    [string] $serviceDir
+)
+
+$cwd = Get-Location
+
+# 0. Find all test directories
+Write-Host "Finding test directories in 'sdk/$serviceDir'"
+$testDirs = & $PSScriptRoot/get_test_dirs.ps1 -serviceDir sdk/$serviceDir
+# Issues here, not returning any objects
+Write-Host "Found test directories $testDirs"
+$temp = Get-Location
+Write-Host "Currently in $temp"
+
+$runTests = $false
+# 0b. Verify there are test files with tests to run in at least one of these directories
+$testDirs | ForEach-Object {
+    Write-Host "[Loop]: Currently in: $_"
+    Get-ChildItem -Path $_ -Filter *_test.go | ForEach-Object {
+        Write-Host "[Loop]: $_"
+        if (Select-String -path $_ -pattern 'Test' -SimpleMatch) {
+            $runTests = $true
+        }
+    }
+
+    if (!$runTests) {
+        Write-Host "There were no test files found."
+        Exit 0
+    }
+}
+
+Write-Host "Proceeding to run tests and add coverage"
+
+# 1. Run tests
+foreach ($td in $testDirs) {
+    Push-Location $td
+    $temp = Get-Location
+    Write-Host "Currently in $temp"
+    Write-Host "##[command] Executing go test -run ""^Test"" -race -v -coverprofile coverage.txt -covermode atomic $td | go-junit-report -set-exit-code > report.xml"
+    go test -run "^Test" -race -v -coverprofile coverage.txt -covermode atomic . | go-junit-report -set-exit-code > report.xml
+    if (!$?) {
+        Write-Host "There was an error running the tests"
+        Exit $LASTEXITCODE
+    } else {
+        Write-Host "" "Successfully ran test suite at $temp" ""
+    }
+    # if no tests were actually run (e.g. examples) delete the coverage file so it's omitted from the coverage report
+    if (Select-String -path ./report.xml -pattern '<testsuites></testsuites>' -simplematch -quiet) {
+        Write-Host "##[command] Deleting empty coverage file"
+        Remove-Item coverage.txt
+        Exit 0
+    }
+}
+
+Set-Location $cwd
+$temp = Get-Location
+Write-Host "Currently in $temp"
+
+# 2. Install coverage tools needed
+go get github.com/jstemmer/go-junit-report
```

Comment:
I wonder if it's worth versioning these (e.g. making coverage its own go mod project directory)? That way we won't by surprised by any dependency changes with these tools that may affect the consistency of our coverage+test report generation.

Reply:
We definitely try and pin versions of other tools to make sure we don't get broken by an unexpected update.